### PR TITLE
API: Add Bfloat16 datatype

### PR DESCRIPTION
**Pull request by lappazos**
## What
Add support to Bfloat16 datatype

## Why ?
Support new DT in CUDA

## How ?
non-reduce collective with size of 16bits, reduce: 
For CPU - convert to FP32, for CUDA - use native CUDA for bfloat16

### CODE REVIEW

**Comment by lappazos** [comment]
@vspetrov we don't need MPI tests for bfloat16 correct? just like FLOAT16?

File: src/utils/ucc_byte_order.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020-2021.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+//TODO was taken from here, rights issue?
+// https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/byte_order.h
```
Comment (manjugv) [inline src/utils/ucc_byte_order.h:7]: Yes. You need to include the Apache license. There is some information on how to do this here:
https://www.apache.org/licenses/LICENSE-2.0

File: src/utils/ucc_math.c [hunk 2]
```diff
@@ -7,19 +7,20 @@
...
-    [UCC_DT_UINT64]  = 8,
-    [UCC_DT_FLOAT64] = 8,
-    [UCC_DT_INT128]  = 16,
-    [UCC_DT_UINT128] = 16,
+    [UCC_DT_INT8]     = 1,
+    [UCC_DT_UINT8]    = 1,
+    [UCC_DT_INT16]    = 2,
+    [UCC_DT_UINT16]   = 2,
+    [UCC_DT_FLOAT16]  = 2,
+	[UCC_DT_BFLOAT16] = 2,
```
Comment (manjugv) [inline src/utils/ucc_math.c:15]: Format.

File: src/utils/ucc_byte_order.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+#ifndef UCC_BYTE_ORDER_H_
+#define UCC_BYTE_ORDER_H_
+
+// Byte order defines provided by gcc. MSVC doesn't define those so
+// we define them here.
+// We assume that all windows platform out there are little endian.
+#if defined(_MSC_VER) && !defined(__clang__)
```
Comment (vspetrov) [inline src/utils/ucc_byte_order.h:15]: What this really do? adds some defins for MS Visual Studio compiler? @manjugv are we even supporting MS ?
Regardless, we don't need that header at all. Instead what we need to do is to detect ENDIANESS during configure step. There is a autotool way to do that:
https://www.gnu.org/software/autoconf/manual/autoconf-2.65/html_node/C-Compiler.html

AC_C_BIGENDIAN

Autotool will run the check during configure, we will define macro in config.h (UCC_BIG_ENDIAN vs UCC_LITTLE_ENDIAN) and use those macros in the code.

File: src/components/mc/cpu/reduce/mc_cpu_reduce.h [hunk 4]
```diff
@@ -162,6 +162,115 @@
...
+        bfloat16tofloat32(&_s2[_i + 5 * _sc]))
+#define OP_BFLOAT16_7(_s1, _s2, _i, _sc, _OP)                                  \
+    _OP((OP_BFLOAT16_6(_s1, _s2, _i, _sc, _OP)),                               \
+        bfloat16tofloat32(&_s2[_i + 6 * _sc]))
+
+#define OP_BFLOAT16_N(_d, _s1, _s2, _sc, _OP, _n)                              \
+    do {                                                                       \
+        size_t _i;                                                             \
+        for (_i = 0; _i < count; _i++) {                                       \
+            float32tobfloat16(OP_BFLOAT16_##_n(_s1, _s2, _i, _sc, _OP),        \
```
Comment (vspetrov) [inline src/components/mc/cpu/reduce/mc_cpu_reduce.h:190]: is this working ? should not there be an assignment d[_i] = float32tobfloat16(...) ?
Did you run gtest/mpitest ? they should have failed. 

Aside: can we do w/o that much code duplication?

**Review by vspetrov — CHANGES_REQUESTED** [review]

File: tools/perf/ucc_pt_config.cc [hunk 5]
```diff
@@ -46,13 +46,14 @@ const std::map<std::string, ucc_datatype_t> ucc_pt_datatype_map = {
...
     {"float16", UCC_DT_FLOAT16},
+    {"bfloat16", UCC_DT_BFLOAT16},
     {"int32", UCC_DT_INT32},
     {"float32", UCC_DT_FLOAT32},
     {"int64", UCC_DT_INT64},
     {"uint64", UCC_DT_UINT64},
     {"float64", UCC_DT_FLOAT64},
     {"int128", UCC_DT_INT128},
-    {"uint128", UCC_DT_UINT128},
+	{"uint128", UCC_DT_UINT128},
```
Comment (lappazos) [inline tools/perf/ucc_pt_config.cc:56]: aligment

**Comment by lappazos** [comment]
NCCL TL Part will added as a separate commit while we get all answers from CUDA team

**Comment by lappazos** [comment]
@vspetrov still needs to test on Selene, where we have the native GPU support for Bfloat16 dt.

**Comment by lappazos** [comment]
@vspetrov @Sergei-Lebedev @manjugv please review

File: test/gtest/core/test_reduce.cc [hunk 6]
```diff
@@ -44,6 +44,7 @@ class test_reduce : public UccCollArgs, public testing::Test {
                 /* need to limit the init value so that "prod" operation
                    would not grow too large. We have teams up to 16 procs
                    in gtest, this would result in prod ~2**48 */
+                /* bFloat16 will be assigned with the floats matching the uint16 bit pattern*/
```
Comment (lappazos) [inline test/gtest/core/test_reduce.cc:47]: change uint16 -> unsigned short

File: test/gtest/core/test_mc_reduce.h [hunk 7]
```diff
@@ -216,11 +242,12 @@ class test_mc_reduce : public testing::Test {
             res_h[i] = (typename T::type)(0);
         }
         for (int i = 0; i < COUNT; i++) {
+            /* bFloat16 will be assigned with the floats matching the uint16 bit pattern*/
```
Comment (lappazos) [inline test/gtest/core/test_mc_reduce.h:245]: change uint16 -> unsigned short

File: test/gtest/core/test_allreduce.cc [hunk 8]
```diff
@@ -41,6 +41,7 @@ class test_allreduce : public UccCollArgs, public testing::Test {
                 /* need to limit the init value so that "prod" operation
                    would not grow too large. We have teams up to 16 procs
                    in gtest, this would result in prod ~2**48 */
+                /* bFloat16 will be assigned with the floats matching the uint16 bit pattern*/
```
Comment (lappazos) [inline test/gtest/core/test_allreduce.cc:44]: change uint16 -> unsigned short

File: src/utils/ucc_math.h [hunk 9]
```diff
@@ -66,4 +66,27 @@ int ucc_sort_uniq(int *array, int len, int inverse);
...
+    ((unsigned short *)(&res))[1] = *((unsigned short *)bfloat16_ptr);
+#endif
+    return res;
+}
+
+static inline void float32tobfloat16(float float_val, void *bfloat16_ptr)
+{
+#if UCC_BIG_ENDIAN
+    *((unsigned short *)bfloat16_ptr) =
+            ((unsigned short *)(&float_val))[0];
```
Comment (Sergei-Lebedev) [inline src/utils/ucc_math.h:85]: why not use uint16_t?

File: src/components/tl/nccl/tl_nccl.h [hunk 10]
```diff
@@ -13,6 +13,12 @@
 #include "components/tl/ucc_tl_log.h"
 #include "utils/ucc_mpool.h"
 
+#if CUDART_VERSION >= 11000 && (NCCL_VERSION_CODE >= NCCL_VERSION(2,10,3))
```
Comment (vspetrov) [inline src/components/tl/nccl/tl_nccl.h:16]: @lappazos this does not compile (i guess you haven't try building with NCCL). you use macro NCCL_VERSION before including nccl.h so it is not defined at line 16. Hence compilation fails. plz remove #if at line 16 and just keep BF16.

**Review by vspetrov — CHANGES_REQUESTED** [review]

File: src/utils/ucc_math.h (hunk 9 above)
Comment (lappazos) [reply src/utils/ucc_math.h:85]: @Sergei-Lebedev @vspetrov 
They did it in CUDA, in cuda_bf16.hpp... you think it's a mistake?

typedef struct __CUDA_ALIGN__(2) {
    unsigned short x;
} __nv_bfloat16_raw;

Comment (lappazos) [reply src/utils/ucc_math.h:85]: from CUDA slack channel - "It is 2 bytes on all targets that CUDA currently supports"

File: src/components/tl/sharp/tl_sharp_coll.c [hunk 11]
```diff
@@ -27,6 +27,7 @@ int ucc_to_sharp_dtype[] = {
     [UCC_DT_FLOAT16]     = SHARP_DTYPE_FLOAT_SHORT,
     [UCC_DT_FLOAT32]     = SHARP_DTYPE_FLOAT,
     [UCC_DT_FLOAT64]     = SHARP_DTYPE_DOUBLE,
+    // TODO in hpcx-2.11 add UCC_DT_BFLOAT16
```
[the remark this answers is not in the document]
Comment (lappazos) [reply src/components/tl/sharp/tl_sharp_coll.c:30]: @Sergei-Lebedev you meant SHARP_DTYPE_NULL right?

File: src/components/tl/nccl/tl_nccl.h (hunk 10 above)
Comment (lappazos) [reply src/components/tl/nccl/tl_nccl.h:16]: @vspetrov i'm keeping the if on CUDA though, right?

Comment (vspetrov) [reply src/components/tl/nccl/tl_nccl.h]: i'm not sure it does any difference but you can keep CUDART_VERSION just for indication that it works for that cuda ver.

File: test/mpi/mpi_util.h [hunk 12]
```diff
@@ -34,6 +34,7 @@ static inline MPI_Datatype ucc_dt_to_mpi(ucc_datatype_t dt) {
     case UCC_DT_FLOAT16:
     case UCC_DT_INT128:
     case UCC_DT_UINT128:
+    case UCC_DT_BFLOAT16:
```
Comment (vspetrov) [inline test/mpi/mpi_util.h:37]: @lappazos plz advice you you are running ucc_test_mpi with bfloat16/float16? Why i try to run it, it get "Unsupported dt".

File: test/gtest/core/test_mc_reduce.h [hunk 13]
```diff
@@ -86,6 +87,31 @@ struct ReductionTest<UCC_DT_FLOAT64, op>
...
+    const static ucc_datatype_t     dt    = UCC_DT_BFLOAT16;
+    const static ucc_reduction_op_t redop = op<float>::redop;
+    static void                     assert_equal(type arg1, type arg2)
+    {
+        // near because of different calculation methods - in CPU op across all
+        // vectors as fp32, and only then convert to bfloat16. here conversion to fp32
+        // is per couple. In CUDA, op is as "real" bfloat16 op.
+        // For example - 15*29=435, which doesn't fit in bfloat16.
+        // casting from fp32 will result in 434, while in GPU we will get 436.
+        ASSERT_NEAR(bfloat16tofloat32(&arg1), bfloat16tofloat32(&arg2),
```
Comment (vspetrov) [inline test/gtest/core/test_mc_reduce.h:102]: @lappazos i don't get it: on the one hand you say that there can be a difference up to 2 (434 vs 436) but adding ASSERT_NEAR with the precision e-33. Is 33 decimal digits intended here?

File: test/gtest/core/test_mc_reduce.h [hunk 14]
```diff
@@ -285,30 +313,55 @@ class test_mc_reduce : public testing::Test {
...
 
-using ReductionTypesOps = ::testing::Types<
-    ReductionTest<UCC_DT_INT16, max>, ReductionTest<UCC_DT_INT32, max>,
-    ReductionTest<UCC_DT_INT64, max>, ReductionTest<UCC_DT_INT16, min>,
-    ReductionTest<UCC_DT_INT32, min>, ReductionTest<UCC_DT_INT64, min>,
-    ReductionTest<UCC_DT_INT16, sum>, ReductionTest<UCC_DT_INT32, sum>,
-    ReductionTest<UCC_DT_INT64, sum>, ReductionTest<UCC_DT_INT16, prod>,
-    ReductionTest<UCC_DT_INT32, prod>, ReductionTest<UCC_DT_INT64, prod>,
-    ReductionTest<UCC_DT_INT16, land>, ReductionTest<UCC_DT_INT32, land>,
-    ReductionTest<UCC_DT_INT64, land>, ReductionTest<UCC_DT_INT16, band>,
```
Comment (vspetrov) [inline test/gtest/core/test_mc_reduce.h:296]: @lappazos why we keep changing this formatting: last time you changed it this way and now  changing back.

File: test/gtest/core/test_mc_reduce_cuda.cc [hunk 15]
```diff
@@ -52,7 +52,11 @@ TYPED_TEST(test_mc_reduce_alpha, ucc_reduce_multi_alpha_cuda) {
         for (int j = 1; j < num_vec; j++) {
             res = TypeParam::do_op(this->buf2_h[i + j * this->COUNT], res);
         }
-        res *= (typename TypeParam::type)alpha;
+        if (TypeParam::dt == UCC_DT_BFLOAT16){
```
Comment (vspetrov) [inline test/gtest/core/test_mc_reduce_cuda.cc:55]: space between ) and {

File: test/gtest/core/test_mc_reduce_cuda.cc [hunk 16]
```diff
@@ -52,7 +52,11 @@ TYPED_TEST(test_mc_reduce_alpha, ucc_reduce_multi_alpha_cuda) {
...
+            float32tobfloat16(bfloat16tofloat32(&res)*(float)alpha, &res);
+        } else{
```
Comment (vspetrov) [inline test/gtest/core/test_mc_reduce_cuda.cc:57]: space after else

File: test/gtest/core/test_mc_reduce_host.cc [hunk 17]
```diff
@@ -46,7 +46,11 @@ TYPED_TEST(test_mc_reduce_alpha, ucc_reduce_multi_alpha_host) {
         for (int j = 1; j < num_vec; j++) {
             res = TypeParam::do_op(this->buf2_h[i + j * this->COUNT], res);
         }
-        res *= (typename TypeParam::type)alpha;
+        if (TypeParam::dt == UCC_DT_BFLOAT16){
+            float32tobfloat16(bfloat16tofloat32(&res)*(float)alpha, &res);
+        } else{
```
Comment (vspetrov) [inline test/gtest/core/test_mc_reduce_host.cc:51]: space after else

Comment (vspetrov) [reply test/gtest/core/test_mc_reduce_host.cc:51]: all over: plz check ) and { have space in between

**Review by vspetrov — APPROVED** [review]

File: test/gtest/core/test_mc_reduce.h (hunk 13 above)
Comment (lappazos) [reply test/gtest/core/test_mc_reduce.h:102]: So the difference isn't absolute, it depends on the range of values were dealing with - the example I gave is regarding round numbers, where the difference is 2 in that specific case. In the range of values we are using in the test (small floats), this difference out here is enough.

File: test/gtest/core/test_mc_reduce.h (hunk 14 above)
Comment (lappazos) [reply test/gtest/core/test_mc_reduce.h:296]: last time was a mistake, i fixed it here

File: src/components/mc/cuda/kernel/mc_cuda_reduce_ops.h [hunk 18]
```diff
@@ -21,4 +24,22 @@
 #define DO_OP_PROD_HALF(_v1, _v2) (__float2half(__half2float(_v1) * __half2float(_v2)))
 #endif
 
+#if defined(__CUDA_BF16_TYPES_EXIST__)
```
Comment (Sergei-Lebedev) [inline src/components/mc/cuda/kernel/mc_cuda_reduce_ops.h:27]: I think we agreed that CUDART_VERSION >= 11000 is equivalent to  __CUDA_BF16_TYPES_EXIST__, let's use only CUDART_VERSION >= 11000 everywhere

File: test/gtest/core/test_allreduce.cc [hunk 19]
```diff
@@ -127,7 +129,11 @@ class test_allreduce : public UccCollArgs, public testing::Test {
                 res = T::do_op(res, ((typename T::type *)((ctxs[r])->init_buf))[i]);
             }
             if (T::redop == UCC_OP_AVG) {
-                res = res / (typename T::type)ctxs.size();
+            	if (T::dt == UCC_DT_BFLOAT16){
+            		float32tobfloat16(bfloat16tofloat32(&res)/(float)ctxs.size(), &res);
```
Comment (Sergei-Lebedev) [inline test/gtest/core/test_allreduce.cc:133]: alignment

File: test/gtest/core/test_mc_reduce_cuda.cc [hunk 20]
```diff
@@ -52,7 +52,11 @@ TYPED_TEST(test_mc_reduce_alpha, ucc_reduce_multi_alpha_cuda) {
         for (int j = 1; j < num_vec; j++) {
             res = TypeParam::do_op(this->buf2_h[i + j * this->COUNT], res);
         }
-        res *= (typename TypeParam::type)alpha;
+        if (TypeParam::dt == UCC_DT_BFLOAT16) {
+            float32tobfloat16(bfloat16tofloat32(&res)*(float)alpha, &res);
+        } else {
+        	res *= (typename TypeParam::type)alpha;
```
Comment (Sergei-Lebedev) [inline test/gtest/core/test_mc_reduce_cuda.cc:58]: alignment

File: test/gtest/core/test_mc_reduce_host.cc [hunk 21]
```diff
@@ -46,7 +46,11 @@ TYPED_TEST(test_mc_reduce_alpha, ucc_reduce_multi_alpha_host) {
         for (int j = 1; j < num_vec; j++) {
             res = TypeParam::do_op(this->buf2_h[i + j * this->COUNT], res);
         }
-        res *= (typename TypeParam::type)alpha;
+        if (TypeParam::dt == UCC_DT_BFLOAT16) {
+            float32tobfloat16(bfloat16tofloat32(&res)*(float)alpha, &res);
+        } else {
+        	res *= (typename TypeParam::type)alpha;
```
Comment (Sergei-Lebedev) [inline test/gtest/core/test_mc_reduce_host.cc:52]: alignment

File: test/gtest/core/test_reduce.cc [hunk 22]
```diff
@@ -126,7 +128,11 @@ class test_reduce : public UccCollArgs, public testing::Test {
                               ((typename T::type *)((ctxs[r])->init_buf))[i]);
             }
             if (T::redop == UCC_OP_AVG) {
-                res = res / (typename T::type)ctxs.size();
+            	if (T::dt == UCC_DT_BFLOAT16){
+            		float32tobfloat16(bfloat16tofloat32(&res)/(float)ctxs.size(), &res);
```
Comment (Sergei-Lebedev) [inline test/gtest/core/test_reduce.cc:132]: alignment

File: test/gtest/core/test_mc_reduce_host.cc [hunk 23]
```diff
@@ -65,7 +69,11 @@ TYPED_TEST(test_mc_reduce_alpha, ucc_reduce_multi_alpha_host_many_vectors) {
         for (int j = 1; j < num_vec; j++) {
             res = TypeParam::do_op(this->buf2_h[i + j * this->COUNT], res);
         }
-        res *= (typename TypeParam::type)alpha;
+        if (TypeParam::dt == UCC_DT_BFLOAT16){
+            float32tobfloat16(bfloat16tofloat32(&res)*(float)alpha, &res);
+        } else{
+        	res *= (typename TypeParam::type)alpha;
```
Comment (Sergei-Lebedev) [inline test/gtest/core/test_mc_reduce_host.cc:75]: alignment

**Review by manjugv — APPROVED** [review]

**Review by Sergei-Lebedev — APPROVED** [review]

**Pull request merged**